infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -0,0 +1,332 @@
+     * expected epoch.
+     * 3. Other Exception is thrown. We do nothing. (The response[i] in this case is left behind.)
+     *
+     * @param pollIteration poll iteration in the ongoing polling round.
+     */
+    private void collectResponsesAndVerifyEpochs(int pollIteration) {
+        // Collect responses and increment response counters for successful pings.
+        for (int i = 0; i < members.length; i++) {
+            try {
+                pollCompletableFutures[i].get();
+                responses[i] = pollIteration;
+                expectedEpoch[i] = INVALID_EPOCH;
+            } catch (Exception e) {
+                if (e.getCause() instanceof WrongEpochException) {
+                    responses[i] = pollIteration;
+                    expectedEpoch[i] = ((WrongEpochException) e.getCause()).getCorrectEpoch();
+                }
+            }
+        }
+    }
+
+    /**
+     * This method performs 2 tasks:
+     * 1. Aggregates the successful responses in the responsiveNodes Set.
+     * If the response was unsuccessful we increment the timeout period.
+     * 2. Checks if there were no failures and no wrongEpochExceptions in which case the ongoing
+     * polling round is completed. However, if failures were present (not epoch errors), then all
+     * the routers' response timeouts are updated with the increased new period and the round
+     * continues.
+     *
+     * @param membersSet    Set of member servers.
+     * @param pollIteration Iteration of the ongoing polling round.
+     * @return True if no failures were present. False otherwise.
+     */
+    private boolean isFailurePresent(Set<String> membersSet, int pollIteration) {
+
+        // Aggregate the responses.
+        Set<String> responsiveNodes = new HashSet<>();
+        long newPeriod = 0;
+        for (int i = 0; i < members.length; i++) {
+            // If this counter is left behind and this node is not in members
+            if (membersSet.contains(members[i]) && responses[i] != pollIteration) {
+                // The existing router response timeout is increased.
+                newPeriod = getIncreasedPeriod();
+            } else {
+                responsiveNodes.add(members[i]);
+            }
+        }
+
+        // We received responses from all the members.
+        if (responsiveNodes.equals(membersSet)) {
+            // We can try to scale back the network latency time only if all pings received.
+            period = Math.max(initPeriodDuration, (period - periodDelta));
+            // If there are no ping failures and no out of phase epochs then end round.
+            if (Arrays.stream(expectedEpoch).filter(l -> l != INVALID_EPOCH).count() == 0) {
+                return false;
+            }
+
+        } else {
+            if (newPeriod != period && newPeriod != 0) {
+                period = newPeriod;
+                tuneRoutersResponseTimeout(membersSet, period);
```

Comment:
this method should not have a side effect like tuningRoutersResponseTimeout .

Reply:
Removed side effect